File: modules/client-java/src/test/java/be/wegenenverkeer/atom/java/FeedProcessorTest.java
package be.wegenenverkeer.atom.java;

import be.wegenenverkeer.atom.EntryRef;
import be.wegenenverkeer.atom.FeedProcessingException;
import be.wegenenverkeer.atom.Url;
import org.joda.time.LocalDateTime;
import org.junit.Test;
import scala.Option;

import java.util.ArrayList;
import java.util.List;

public class FeedProcessorTest {

    private static final String FEED_URL = "http://example.com/feeds/";
    private static final String FEED_URL_PAGE1 = FEED_URL + "1";

    @Test
    public void test() {
        System.out.println("Start processing");

        // create the feed provider
        ExampleFeedProvider provider = new ExampleFeedProvider();

        // create your entry consumer
        ExampleEntryConsumer consumer = new ExampleEntryConsumer();

        // create the processor
        be.wegenenverkeer.atom.java.FeedProcessor<ExampleFeedEntry> processor = new be.wegenenverkeer.atom.java.FeedProcessor<ExampleFeedEntry>(provider, consumer);

        // start processing the new feed pages and its entries
        processor.start();
    }

    static class ExampleFeedEntry {
        private static int COUNTER;
        private int counter;

        ExampleFeedEntry() {
            this.counter = COUNTER++;
        }

        @Override
        public String toString() {
            return "ExampleFeedEntry{" +
                    "counter=" + counter +
                    '}';
        }
    }

    static class ExampleEntryConsumer implements be.wegenenverkeer.atom.java.EntryConsumer<ExampleFeedEntry> {
        @Override
        public Entry<ExampleFeedEntry> accept(Entry<ExampleFeedEntry> entry) {
            System.out.println("Consuming entry " + entry.getContent());
            try {
                handleEvent(entry.getContent().getValue());
                return entry;
            } catch (Exception e) {
                throw new FeedProcessingException(Option.apply(entry.getId()), e.getMessage());
            }
        }

        public void handleEvent(ExampleFeedEntry event) throws Exception {
            // handle the new event here and persist the current feed position here (possibly in 1 database transaction)
            System.out.println("process feed entry and persist feed position " + event);
        }
    }

    static class ExampleFeedProvider implements be.wegenenverkeer.atom.java.FeedProvider<ExampleFeedEntry> {

        /**
         * @return the EntryRef from where you want to start processing
         * A null EntryRef assure that all items will be processed.
         */
        @Override
        public EntryRef<ExampleFeedEntry> getInitialEntryRef() {
            return null;
        }

        @Override
        public Feed<ExampleFeedEntry> fetchFeed() {
            return fetchFeed(FEED_URL_PAGE1);
        }

        @Override
        public Feed<ExampleFeedEntry> fetchFeed(String page) {
            System.out.println("Fetching page " + page);
            int intPage = 0;
            //intPage = Integer.parseInt(page);  // @todo make it fail --- with message "null"
            if (page.endsWith("/1")) intPage=1;
            if (page.endsWith("/2")) intPage=2;
            if (page.endsWith("/3")) intPage=3;

            List<Entry<ExampleFeedEntry>> entries = new ArrayList<Entry<ExampleFeedEntry>>();
            List<Link> links = new ArrayList<Link>();
            entries.add(new Entry<ExampleFeedEntry>("id1", new Content<ExampleFeedEntry>(new ExampleFeedEntry(), ""), links));
            entries.add(new Entry<ExampleFeedEntry>("id2", new Content<ExampleFeedEntry>(new ExampleFeedEntry(), ""), links));
            entries.add(new Entry<ExampleFeedEntry>("id3", new Content<ExampleFeedEntry>(new ExampleFeedEntry(), ""), links));

            List<Link> feedLinks = new ArrayList<Link>();
            feedLinks.add(new Link("last", FEED_URL_PAGE1));
            feedLinks.add(new Link("self", FEED_URL + intPage));
            if (intPage < 3) {
                feedLinks.add(new Link("previous", FEED_URL + (intPage + 1)));
            }
            if (intPage > 0) {
                feedLinks.add(new Link("next", FEED_URL + (intPage - 1)));
            }

            Feed<ExampleFeedEntry> feed = new Feed<ExampleFeedEntry>(
                    "id",
                    FEED_URL,
                    "Blabla",
                    null,
                    new LocalDateTime(),
                feedLinks,
                    entries
            );
            return feed;
        }

    }
}

File: modules/client-scala/src/test/scala/be/wegenenverkeer/atom/FeedProcessorTest.scala
package be.wegenenverkeer.atom

import org.joda.time.LocalDateTime
import org.scalatest.{FlatSpec, Matchers}

import scala.annotation.tailrec
import scala.util.{Failure, Success, Try}


class FeedProcessorTest extends FlatSpec with Matchers {

  type StringFeed = Feed[String]
  type Feeds = List[StringFeed]

  case class Scenario(provider:TestFeedProvider,
                      consumedEvents:List[String]) {

    val consumer = new StatefulEntryConsumer
    val process = new FeedProcessor[String](provider, consumer)

    val result = Try(process.start())

    consumedEvents shouldBe consumer.consumedEvents.toList
    lastConsumedEntryId shouldBe consumer.lastConsumedEntryId

    def lastConsumedEntryId:Option[String] = {
      consumedEvents.reverse.headOption
    }

    def assertResult(block:  Try[AtomResult[String]] => Unit) = block(result)

    def shouldSucceed = {

      assertResult { result =>
        result.flatMap(_.asTry).isSuccess shouldBe true
      }

      new {
        def consumingNothing: this.type = {
          result match {
            case Success(AtomNothing) => this
            case anyOther => fail(s"Expecting a AtomNothing, got a $anyOther")
          }
        }
        def lastConsuming(lastConsumedId:String): this.type = {
          result match {
            case Success(AtomSuccess(Some(entry))) => entry.id shouldBe lastConsumedId
            case anyOther => fail(s"Expecting a AtomSuccess($lastConsumedId), got a $anyOther")
          }
          this
        }
      }
    }


    def shouldFail : this.type = {
      assertResult { result =>
        result.flatMap(_.asTry).isFailure shouldBe true
      }
      this
    }
  }


  "FeedProcessor" should "NOT terminate with an error if feed is empty" in {
    Scenario(
      provider = feedProvider(None),
      consumedEvents = List()
    ).shouldSucceed.consumingNothing
  }

  it should "process all entries when starting without an initial entry id" in {
    Scenario(
      provider = feedProvider(
        startingFrom = None,
        feed("feed/0/3")("a1", "b1", "c1"),
        feed("feed/3/3")("a2", "b2", "c2"),
        feed("feed/6/3")("a3", "b3", "c3")
      ),
    
      consumedEvents = List("a1", "b1", "c1", "a2", "b2", "c2", "a3", "b3", "c3")
    ).shouldSucceed.lastConsuming("c3")
  }

  it should "process all remaining entries when starting in the middle of a page" in {
    Scenario(
      provider = feedProvider(
        startingFrom = pos("http://www.example.org/feeds/feed/3/3", "b2"),
        feed("feed/0/3")("a1", "b1", "c1"),
        feed("feed/3/3")("a2", "b2", "c2"),
        feed("feed/6/3")("a3", "b3", "c3")
      ),

      consumedEvents = List("c2", "a3", "b3", "c3")
    ).shouldSucceed.lastConsuming("c3")
  }

  it should "process all remaining entries when starting a new page" in {
    Scenario(
      provider = feedProvider(
        startingFrom = pos("http://www.example.org/feeds/feed/3/3", "c2"),
        feed("feed/0/3")("a1", "b1", "c1"),
        feed("feed/3/3")("a2", "b2", "c2"),
        feed("feed/6/3")("a3", "b3", "c3")
      ),

      consumedEvents = List("a3", "b3", "c3")
    ).shouldSucceed.lastConsuming("c3")
  }

  it should "terminate with an error when started with a non-existent entry" in {
    Scenario(
      provider = feedProvider(
        startingFrom = pos("http://www.example.org/feeds/feed/0/3", "d1"),
        feed("feed/0/3")("a1", "b1", "c1"),
        feed("feed/3/3")("a2", "b2", "c2"),
        feed("feed/6/3")("a3", "b3", "c3")
      ),

      consumedEvents = List()
    ).shouldFail
  }

  it should "process nothing when starting from the last entry" in {
    Scenario(
      provider = feedProvider(
        startingFrom = pos("http://www.example.org/feeds/feed/6/3", "c3"),
        feed("feed/0/3")("a1", "b1", "c1"),
        feed("feed/3/3")("a2", "b2", "c2"),
        feed("feed/6/3")("a3", "b3", "c3")
      ),

      consumedEvents = List()
    ).shouldSucceed.consumingNothing
  }

  it should "return a Failure if provider throws an exception" in {

    Scenario(
      provider = feedProviderBogus(
        feed("feed/0/3")("a1", "b1", "c1"),
        feed("feed/3/3")("a2", "b2", "c2"),
        feed("feed/6/3")("a3", "b3", "c3")
      ),

      consumedEvents = List("a1", "b1", "c1")

    ).shouldFail
  }

  it should "return a Failure if consumer throws an exception" in {
    val provider = feedProvider(
      startingFrom = None,
      feed("feed/0/3")("a1", "b1", "c1"),
      feed("feed/3/3")("a2", "b2", "c2"),
      feed("feed/6/3")("a3", "b3", "c3")
    )

    val errorMessage = "Error when consuming Entry"
    val consumer = new EntryConsumer[String] {
      override def apply(eventEntry: Entry[String]): Try[Entry[String]] = {
        throw new RuntimeException(errorMessage)
      }
    }

    val processor = new FeedProcessor[String](provider, consumer)
    val result = processor.start()
    result.asTry.isFailure shouldBe true
    result.asTry.failed.map { error =>
      error shouldBe errorMessage
    }
  }



  def feed(url:String)(events:String*) : Feed[String] = {
    val entries = events.map { e =>
      val content = Content[String](e, "")
      Entry[String](e, new LocalDateTime(), content, Nil)
    }

    val links = List(Link(Link.selfLink, Url("http://www.example.org/feeds") / url))

    Feed(
      id = "id",
      base = Url("http://www.example.org/feeds"),
      title = Option("title"),
      generator = None,
      updated = new LocalDateTime(),
      links = links,
      entries = entries.toList
    )
  }

  def pos(url:String, entryId: String) : Option[EntryRef[String]] = {
    Some(EntryRef[String](entryId, Url(url)))
  }


  def feedProvider(startingFrom:Option[EntryRef[String]],
                   feeds:Feed[String]*) = new TestFeedProvider(startingFrom, feeds.toList)

  /**
   * Bogus provider. Never returns the next Feed
   */
  def feedProviderBogus(feeds:Feed[String]*) = new TestFeedProvider(None, feeds.toList) {
    override def fetchFeed(page: String): Try[Feed[String]] = {
      Failure(FeedProcessingException(None, "Can't fetch feed"))
    }
  }

  class TestFeedProvider(val initialEntryRef:Option[EntryRef[String]], feeds:List[Feed[String]]) extends FeedProvider[String] {

    val linkedFeeds = {
      // build links between feeds
      feeds match {
        case x :: xs => linkFeeds(x.selfLink, x, xs)
        case Nil => List()
      }
    }

    /**
     * add 'last' and 'previous' link to list of Feeds
     */
    @tailrec
    private def linkFeeds(lastLink:Link, current:StringFeed, others: Feeds, acc:Feeds = List()) : Feeds = {

      def addLink(feed:StringFeed, linkLabel:String, link:Link) : StringFeed = {
        feed.copy (
          links = feed.links :+ link.copy(rel = linkLabel)
        )
      }

      // last link must always be present
      val currentWithLastLink = addLink(current, Link.lastLink, lastLink)

      others match {
        case Nil => acc :+ currentWithLastLink
        case x :: xs =>
          val currentWithPreviousLink = addLink(currentWithLastLink, Link.previousLink, x.selfLink)
          linkFeeds(lastLink, x, xs, acc :+ currentWithPreviousLink)
      }

    }

    /**
     * Return first feed or a Failure
     */
    override def fetchFeed(): Try[Feed[String]] = {
      initialEntryRef match {
        case None => optToTry(linkedFeeds.headOption)
        case Some(position) => fetchFeed(position.url.path)
      }

    }

    /**
     * Return feed whose selfLink equals 'page or Failure
     */
    override def fetchFeed(page: String): Try[Feed[String]] = {
      val feedOpt =  linkedFeeds.find {
        feed => feed.resolveUrl(feed.selfLink.href).path == page
      }
      optToTry(feedOpt)
    }

    private def optToTry(feedOpt:Option[Feed[String]]): Success[Feed[String]] = {

      def emptyFeed = {
        val links = List(Link(Link.selfLink, Url("http://www.example.org/feeds")))
        Feed(
          id = "id",
          base = Url("http://www.example.org/feeds"),
          title = Option("title"),
          generator = None,
          updated = new LocalDateTime(),
          links = links,
          entries = List()
        )
      }
      feedOpt.map { feed =>
        Success(feed)
      }.getOrElse(Success(emptyFeed))
    }

  }

}
